In [1]:
import tensorflow as tf
from tensorflow import keras
import time

# Check GPU availability
print("GPU available:", tf.config.list_physical_devices('GPU'))

# Load MNIST dataset
(x_train, y_train), (x_test, y_test) = keras.datasets.mnist.load_data()
x_train = x_train.reshape(-1, 28, 28, 1).astype('float32') / 255.0
x_test = x_test.reshape(-1, 28, 28, 1).astype('float32') / 255.0

# Convert labels to one-hot encoding
y_train = keras.utils.to_categorical(y_train, 10)
y_test = keras.utils.to_categorical(y_test, 10)

GPU available: [PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]
11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step


In [2]:
def create_model():
    model = keras.Sequential([
        keras.layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
        keras.layers.MaxPooling2D((2, 2)),
        keras.layers.Conv2D(64, (3, 3), activation='relu'),
        keras.layers.MaxPooling2D((2, 2)),
        keras.layers.Flatten(),
        keras.layers.Dense(128, activation='relu'),
        keras.layers.Dense(10, activation='softmax')
    ])
    model.compile(optimizer='adam',
                  loss='categorical_crossentropy',
                  metrics=['accuracy'])
    return model

In [3]:
# Force TensorFlow to use CPU
with tf.device('/CPU:0'):
    # Create model
    cpu_model = create_model()

    # Train and time
    start_time = time.time()
    cpu_model.fit(x_train, y_train, epochs=5, batch_size=64, validation_split=0.2)
    cpu_time = time.time() - start_time

    print(f"\n✅ CPU Training completed in {cpu_time:.2f} seconds")

/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/5
750/750 ━━━━━━━━━━━━━━━━━━━━ 40s 51ms/step - accuracy: 0.8770 - loss: 0.3982 - val_accuracy: 0.9816 - val_loss: 0.0662
Epoch 2/5
750/750 ━━━━━━━━━━━━━━━━━━━━ 40s 50ms/step - accuracy: 0.9829 - loss: 0.0548 - val_accuracy: 0.9835 - val_loss: 0.0531
Epoch 3/5
750/750 ━━━━━━━━━━━━━━━━━━━━ 38s 50ms/step - accuracy: 0.9893 - loss: 0.0360 - val_accuracy: 0.9872 - val_loss: 0.0443
Epoch 4/5
750/750 ━━━━━━━━━━━━━━━━━━━━ 41s 51ms/step - accuracy: 0.9919 - loss: 0.0263 - val_accuracy: 0.9883 - val_loss: 0.0397
Epoch 5/5
750/750 ━━━━━━━━━━━━━━━━━━━━ 40s 50ms/step - accuracy: 0.9943 - loss: 0.0175 - val_accuracy: 0.9852 - val_loss: 0.0495

✅ CPU Training completed in 199.74 seconds


In [3]:
# Allow TensorFlow to use GPU (default behavior)
gpu_model = create_model()

# Train and time
start_time = time.time()
gpu_model.fit(x_train, y_train, epochs=5, batch_size=64, validation_split=0.2)
gpu_time = time.time() - start_time

print(f"\n✅ GPU Training completed in {gpu_time:.2f} seconds")

/usr/local/lib/python3.11/dist-packages/keras/src/layers/convolutional/base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/5
750/750 ━━━━━━━━━━━━━━━━━━━━ 8s 5ms/step - accuracy: 0.8775 - loss: 0.4065 - val_accuracy: 0.9798 - val_loss: 0.0695
Epoch 2/5
750/750 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9814 - loss: 0.0582 - val_accuracy: 0.9859 - val_loss: 0.0472
Epoch 3/5
750/750 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.9897 - loss: 0.0364 - val_accuracy: 0.9881 - val_loss: 0.0384
Epoch 4/5
750/750 ━━━━━━━━━━━━━━━━━━━━ 5s 5ms/step - accuracy: 0.9919 - loss: 0.0247 - val_accuracy: 0.9883 - val_loss: 0.0393
Epoch 5/5
750/750 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.9935 - loss: 0.0188 - val_accuracy: 0.9889 - val_loss: 0.0402

✅ GPU Training completed in 26.79 seconds


CPU completed training in 199.74 sec AND
GPU cpmpleted training in 26.79 sec (GPU is much faster)